File: src/main/java/com/company/erp/core/database/PageQuery.java
package com.company.erp.core.database;

import com.company.erp.core.exceptions.BadRequestException;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;

import java.util.Set;

/**
 * Shared query convention for list endpoints:
 * {@code ?page=1&pageSize=20&search=…&sort=field:asc}.
 *
 * <ul>
 *   <li>{@code page} is 1-indexed (translated to Spring's 0-based internally).</li>
 *   <li>{@code pageSize} is clamped to {@code 1..MAX_PAGE_SIZE}.</li>
 *   <li>{@code sort} is {@code field:asc|desc}; controllers pass an explicit
 *       whitelist of safe fields.</li>
 * </ul>
 */
public record PageQuery(int page, int pageSize, String search, String sort) {

    public static final int MAX_PAGE_SIZE = 100;

    public PageQuery {
        if (page <= 0) page = 1;
        if (pageSize <= 0) pageSize = 20;
        if (pageSize > MAX_PAGE_SIZE) pageSize = MAX_PAGE_SIZE;
    }

    public Pageable toPageable(Set<String> allowedSortFields, Sort defaultSort) {
        Sort resolved = (sort == null || sort.isBlank())
                ? defaultSort
                : parseSort(sort, allowedSortFields);
        return PageRequest.of(page - 1, pageSize, resolved);
    }

    private static Sort parseSort(String raw, Set<String> allowed) {
        String[] parts = raw.split(":");
        String field = parts.length > 0 ? parts[0].trim() : "";
        String dir   = parts.length > 1 ? parts[1].trim().toLowerCase() : "asc";
        if (!allowed.contains(field)) {
            throw new BadRequestException("Cannot sort by '" + field + "'");
        }
        Sort.Direction direction = switch (dir) {
            case "asc"  -> Sort.Direction.ASC;
            case "desc" -> Sort.Direction.DESC;
            default -> throw new BadRequestException("Sort direction must be 'asc' or 'desc'");
        };
        return Sort.by(direction, field);
    }
}

File: app/Support/Pagination/PageQuery.php
<?php

namespace App\Support\Pagination;

use App\Support\Exceptions\BadRequestException;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Http\Request;

/**
 * Shared query convention for list endpoints:
 *   ?page=1&pageSize=20&search=…&sort=field:asc
 *
 * - page is 1-indexed
 * - pageSize is clamped to 1..MAX_PAGE_SIZE
 * - sort is field:asc|desc; callers pass an explicit whitelist of safe fields,
 *   keyed by API field name -> DB column.
 *
 * Port of the Spring PageQuery record.
 */
class PageQuery
{
    const MAX_PAGE_SIZE = 100;

    /** @var int */
    public $page;

    /** @var int */
    public $pageSize;

    /** @var string|null */
    public $search;

    /** @var string|null */
    public $sort;

    public function __construct(int $page, int $pageSize, ?string $search, ?string $sort)
    {
        if ($page <= 0) {
            $page = 1;
        }
        if ($pageSize <= 0) {
            $pageSize = 20;
        }
        if ($pageSize > self::MAX_PAGE_SIZE) {
            $pageSize = self::MAX_PAGE_SIZE;
        }
        $this->page = $page;
        $this->pageSize = $pageSize;
        $this->search = $search;
        $this->sort = $sort;
    }

    public static function fromRequest(Request $request): self
    {
        return new self(
            (int) $request->query('page', 1),
            (int) $request->query('pageSize', 20),
            $request->query('search'),
            $request->query('sort')
        );
    }

    /**
     * Apply the resolved sort to a query builder.
     *
     * @param array<string, string> $allowed  apiField => dbColumn
     * @param array{0:string,1:string} $default  [dbColumn, direction]
     */
    public function applySort(Builder $query, array $allowed, array $default): Builder
    {
        if ($this->sort === null || trim($this->sort) === '') {
            return $query->orderBy($default[0], $default[1]);
        }

        $parts = explode(':', $this->sort);
        $field = isset($parts[0]) ? trim($parts[0]) : '';
        $dir = isset($parts[1]) ? strtolower(trim($parts[1])) : 'asc';

        if (! array_key_exists($field, $allowed)) {
            throw new BadRequestException("Cannot sort by '".$field."'");
        }
        if ($dir !== 'asc' && $dir !== 'desc') {
            throw new BadRequestException("Sort direction must be 'asc' or 'desc'");
        }

        return $query->orderBy($allowed[$field], $dir);
    }
}
